arch/simd_vector.cc

```diff
@@ -541,6 +541,44 @@ double vec_dotprod_sse2(const double * x, const double * y, size_t n)
     return result;
 }
 
+void vec_minus_sse2(const float * x, const float * y, float * r, size_t n)
+{
+    size_t i = 0;
+
+    if (false) ;
+    else {
+        //cerr << "unoptimized" << endl;
+        
+        for (; i + 16 <= n;  i += 16) {
+            v4sf yyyy0 = _mm_loadu_ps(y + i + 0);
+            v4sf xxxx0 = _mm_loadu_ps(x + i + 0);
+            v4sf yyyy1 = _mm_loadu_ps(y + i + 4);
+            xxxx0 -= yyyy0;
+            v4sf xxxx1 = _mm_loadu_ps(x + i + 4);
+            __builtin_ia32_storeups(r + i + 0, xxxx0);
+            v4sf yyyy2 = _mm_loadu_ps(y + i + 8);
+            xxxx1 -= yyyy1;
+            v4sf xxxx2 = _mm_loadu_ps(x + i + 8);
+            __builtin_ia32_storeups(r + i + 4, xxxx1);
+            v4sf yyyy3 = _mm_loadu_ps(y + i + 12);
+            xxxx2 -= yyyy2;
+            v4sf xxxx3 = _mm_loadu_ps(x + i + 12);
+            __builtin_ia32_storeups(r + i + 8, xxxx2);
+            xxxx3 -= yyyy3;
+            __builtin_ia32_storeups(r + i + 12, xxxx3);
+        }
+
+        for (; i + 4 <= n;  i += 4) {
+            v4sf yyyy0 = _mm_loadu_ps(y + i + 0);
+            v4sf xxxx0 = _mm_loadu_ps(x + i + 0);
+            xxxx0 -= yyyy0;
+            __builtin_ia32_storeups(r + i + 0, xxxx0);
+        }
+        
+        for (; i < n;  ++i) r[i] = x[i] - y[i];
```

Comment:
No... this is a typical geometric unrolling strategy.  The first loop is unrolled to 16 values at once, but must have at least 16 to run.  Rather than do the remainder one by one, we then do it 4 by 4, and the remaining 0 to 3 individually.